apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/ClientPool.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax.Grpc;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.Spanner.V1;
+using Google.Cloud.Spanner.V1.Logging;
+using Grpc.Auth;
+
+namespace Google.Cloud.Spanner
+{
+    internal static class ClientPool
+    {
+        private static readonly ConcurrentDictionary<ClientPoolKey, ClientPoolEntry> s_clientEntryPool =
+            new ConcurrentDictionary<ClientPoolKey, ClientPoolEntry>();
+
+        public static int Timeout { get; set; }
+
+        public static Task<SpannerClient> AcquireClientAsync()
+        {
+            return AcquireClientAsync(null, null);
+        }
+
+        public static async Task<SpannerClient> AcquireClientAsync(ITokenAccess credentials, ServiceEndpoint endpoint)
+        {
+            var key = new ClientPoolKey {
+                Credential = credentials ?? ApplicationDefault.Instance,
+                Endpoint = endpoint ?? SpannerClient.DefaultEndpoint
+            };
+            var poolEntry = s_clientEntryPool.GetOrAdd(key, k => new ClientPoolEntry(key));
+            var result = await poolEntry.AcquireClientFromEntryAsync().ConfigureAwait(false);
+            Logger.LogPerformanceCounter("SpannerClient.Count", () => s_clientEntryPool.Count);
+            return result;
+        }
+
+        /// <summary>
+        ///     Callers should be careful not to call AcquireClient while CloseAll is still running.
+        ///     It is possible to obtain a stale client.
+        /// </summary>
+        /// <returns></returns>
+        public static async Task CloseAllAsync()
+        {
+            Logger.Debug(() => "Shutting down all gRPC channels.");
+            await SpannerClient.ShutdownDefaultChannelsAsync().ConfigureAwait(false);
+            s_clientEntryPool.Clear();
+        }
+
+        private struct ClientPoolKey
```

Comment:
Given that you're using this as a dictionary key, I'd definitely implement `IEqualityComparer<ClientPoolKey>`. I'd also use properties rather than public fields - not desperately important at the moment given that it's private, but I just get shivers with public fields...